acp/DPI.md

```diff
@@ -0,0 +1,45 @@
+## DefraDB Policy Interface (DPI) Restrictions
+To be a DPI compliant resource, these rules MUST be met:
+- The resource must exist.
+- The resource must atleast have the required registerer relation under `relations`.
+- The resource must atleast have all the required permissions under `permissions`.
+- Every required permission must have the required registerer relation in `expr`.
+- The required registerer relation must be the leading (first) relation in `expr` (see example below).
+
+To be a 'DPI' policy, all of it's resources must be DPI compliant.
+To be 'Partially-DPI' atleast one of it's resource must be DPI compliant.
+
+Learn more about DefraDB ACP Terminologies [TERMINOLOGY](/acp/TERMINOLOGY.md)
+
+### Moreover:
+All the required permissions are:
+    - Defined in the `dpi.go` file in variable `dpiRequiredPermissions`.
+The required registerer relation name is:
+    - Defined in the `dpi.go` file in variable `requiredRegistererRelationName`.
+
+## DPI Resource Examples:
+- Check out tests here: [tests/integration/acp/schema/add_dpi](/tests/integration/acp/schema/add_dpi)
+- The tests linked are broken into `accept_*_test.go` and `reject_*_test.go` files.
+- Accepted tests document the valid DPIs (as the schema is accepted).
+- Rejected tests document invalid DPIs (as the schema is rejected).
+- There are also some Partially-DPI tests that are both accepted and rejected depending on the resource.
+
+### DPI Resource's Required Permission's Expression:
+Eventhough the following expressions are valid generic policy expressions, they will make a
+DPI compliant resource loose it's DPI status as these expressions are not in accordance to
+our DPI rules defined above. Assuming these `expr` are under a required permission label:
+- `expr: ownerMalicious + owner`
```

Comment:
suggestion: `ownerMalicious` suggests that there is something significant about this string besides the fact that it doesn't exist as a resource (I am guessing, it is not really clear from this section that that is why it is not valid either, and documentation could be improved here too).  I suggest changing `ownerMalicious` to `fshdhjaj` or `notADeclaredResource`.

Reply:
There is very much something suspicious if `owner` is omitted (or not in the beginning) of a required permission within the expr. If validation had a bug and only checked for first 5 characters to ensure "owner" existance then `ownerMalicious` would pass the bug and now have all access an owner might have.

Let me know how you want me to document it better.

Reply:
I can't see anything in the above documentation that suggests the first few chars must be `owner` - can you please document that?  And I'd still suggest changing `ownerMalicious` to `ownerGFHDSg`.

Reply:
`- The required registerer relation must be the leading (first) relation in `expr` (see example below).`

This is what this line was trying to convey by.

Note the first character after the `owner` must be a `[space]` or `+` or `&` or `-`. To avoid this vulnerability because sourcehub allows owner to be specified in any order and any relation name can be in the beginning, example the name "ownerx" according to sourcehub (which is a vulnerability we want to dis-allow as part of DPI)

Reply:
> Note the first character after the owner must be a [space] or + or & or -.

I'm not sure I understand, it must be exactly `ownerFoo`?  If so then can you please meantion that in this point `- The required registerer ...`

Reply:
> > Note the first character after the owner must be a [space] or + or & or -.
> 
> I'm not sure I understand, it must be exactly `ownerFoo`? If so then can you please meantion that in this point `- The required registerer ...`

Why would it be `ownerFoo`? I noted above that the owner must have ` ` `+` `&` or `-` after the "owner".  `ownerFoo`  is therefore invalid expr

Reply:
"ownerNew" is just testing a dangerous possible edge case which should always be detected as invalid `expr`